framework/aster-frame/src/vm/page_table/cursor.rs

```diff
@@ -0,0 +1,318 @@
+// SPDX-License-Identifier: MPL-2.0
+
+use alloc::{boxed::Box, sync::Arc};
+use core::{marker::PhantomData, ops::Range};
+
+use super::{
+    properties::{in_frame_index, nr_entries_per_frame, page_size},
+    MapInfo, MapOp, MapProperty, PageTable, PageTableConstsTrait, PageTableEntryTrait,
+    PageTableError, PageTableFrame, PageTableMode, PtfRef,
+};
+use crate::{
+    sync::SpinLock,
+    vm::{paddr_to_vaddr, Paddr, Vaddr},
+};
+
+/// The cursor for traversal over the page table.
+///
+/// Doing mapping is somewhat like a depth-first search on a tree, except
+/// that we modify the tree while traversing it. We use a stack to simulate
+/// the recursion.
+pub(super) struct PageTableCursor<
```

Comment:
This is a very general comment about the design of the page table cursor.

Could you please have a look at the way how Rust's standard library [implements `CursorMut`](https://github.com/rust-lang/rust/blob/f2c80133931a5e5dc194359c0451bfaed579b48d/library/alloc/src/collections/btree/map.rs#L2857-L2869) for `BTreeMap`? I follow this idea to [implement `CursorMut`](https://github.com/asterinas/xarray/blob/72a4067a65e8f94cfc193f5f19ebc981c1de9de7/src/cursor.rs#L412-L429) for `XArray`. I think it also works for the page table cursor.

For the current design, I think the definition of `PtrRef` looks wired.
```rust
type PtfRef<E, C> = Arc<SpinLock<PageTableFrame<E, C>>>;
```
 - There is no point to have an `Arc<_>`, because you do not really want to share the page table.
 - There is no point to have a `SpinLock<_>`, because you always have exclusive access to the page table.

It looks like a workaround to make the Rust compiler happy, and yes, the Rust compiler will accept the current code, but I do think it's misuse of both `Arc<_>` and `SpinLock<_>`.